h2/src/main/org/h2/mvstore/tx/Transaction.java

```diff
@@ -202,6 +209,12 @@ public void markStatementEnd() {
     void log(int mapId, Object key, VersionedValue oldValue) {
         long currentState = statusAndLogId.getAndIncrement();
         long logId = getLogId(currentState);
+        if (logId > LOG_ID_MASK) {
```

Comment:
There are two problems here.

1. `logId > LOG_ID_MASK` will be always `false` because `state & LOG_ID_MASK > LOG_ID_MASK` is always false. So this check does not work.

2. `logId` is a previous value before the increment operation. If we got something greater than `LOG_ID_MASK` (`0xff_ffff_ffff`) it means that previous invocation of `log()` created this overflow in `statusAndLogId` after increment, from my point of view it's better to avoid this so condition should probably be `>=` here. If so, `0xff_ffff_ffff` will not be a valid log id anymore. Or we can allow `0xff_ffff_ffff` as a last log id, but in this case overflow control bit may have value 1 that should be documented.

Reply:
You are absolutely right, my bad. I was jumping back and forth between two alternatives you've described, and stuck in the middle.
Taking second alternative now: 0xff_ffff_ffff is a last valid value.